Set NU5 testnet reactivation height. (#5868)
zcash/zcash: doc/release-notes.md

Notable changes
===============

Changes to Testnet NU5 Consensus Rules
--------------------------------------

- In order to better support hardware wallets, transparent signature hash construction as
  defined in [ZIP 244](https://zips.z.cash/zip-0244) has been modified to include a hash
  of the serialization of the amounts of all outputs being spent, along with a hash of all
  spent outputs `scriptPubKeys` values, except in the case that the `ANYONECANPAY` flag is
  set. This allows hardware wallet devices to verify the UTXO amounts without having to
  stream all the previous transactions containing the outputs being spent to the device.
  Also as part of these changes, the transparent signature hash digest now commits
  directly, rather than implicitly, to the sighash type, and the sighash type is
  restricted to a fixed set of valid values. The change to ZIP 244 can be seen 
  [here](https://github.com/zcash/zips/commit/ac9dd97f77869b9191dc9592faa3ebf45475c341).
- This release fixes a bug in `v4.6.0` that caused a consensus failure on the Zcash
  testnet at height `1,779,200`. For details see

>   <https://github.com/zcash/zcash/commit/5990853de3e9f04f28b7a4d59ef9e549eb4b4cca>.

Reviewer: Can we make this a link like:

For details see this [commit](https://github.com/zcash/zcash/commit/5990853de3e9f04f28b7a4d59ef9e549eb4b4cca)

Author: We certainly could, but I'm not sure that's particularly more helpful?

- There have been changes to the Halo2 proving system to improve consistency between the
  specification and the implementation, and these may break compatibility. See for example

>   <https://github.com/zcash/halo2/commit/247cd620eeb434e7f86dc8579774b30384ae02b2>.

Reviewer: Same comment regarding a link.

> - There have been numerous changes to the Orchard circuit impementation since `v4.6.0`.

Reviewer: impementation -> implementation.  =)

>   See <https://github.com/zcash/orchard/issues?q=label%3AM-consensus-change-since-0.1.0-beta-1>

Reviewer: Same comment regarding a link.

  for a complete list.
- A potential Faerie Gold vulnerability affecting the previous activation of NU5 on
  testnet and existing since `v4.6.0` has been mitigated.

NU5 Testnet Reactivation
------------------------

To support the aforementioned testnet consensus changes, the following changes are made in
`zcashd v4.7.0`:

- The consensus branch ID for NU5 is changed to `0xC2D6D0B4`.
- The protocol version indicating NU5-aware testnet nodes is set to `170050`.

> - The testnet reactivation height for NU5 is set to **1,840,000**.

Reviewer: Should this be a code escape, like `1,840,000`?

Author: This was copied from how it was written in the 4.5.0 PR.

Testnet nodes that upgrade to `zcashd v4.7.0` prior to block height `1,840,000` will follow
the new testnet network upgrade. Testnet nodes that are running `zcashd v4.6.0-2` at that
height will need to upgrade to `v4.7.0` and then run with `-reindex`.

Mnemonic Recovery Phrases
-------------------------

> The zcashd wallet has been modified to support BIP 39, which describes how to derive the

Author: We use `zcashd` in all lowercase everywhere else.

wallet's HD seed from a mnemonic phrase.  The mnemonic phrase will be generated on load of
the wallet, or the first time the wallet is unlocked, and is available via the
`z_exportwallet` RPC call. All new addresses produced by the wallet are now derived from
this seed using the HD wallet functionality described in ZIP 32 and ZIP 316. For users
upgrading an existing Zcashd wallet, it is recommended that the wallet be backed up prior
to upgrading to the 4.7.0 Zcashd release.

Following the upgrade to 4.7.0, Zcashd will require that the user confirm that they have

> backed up their new emergency recovery phrase, which may be obtained from the output of

Reviewer 2: The preferred user-facing term is "emergency recovery phrase", which is what `zcashd-wallet-tool` says in its user interface. We also use the term "mnemonic seed" which is more jargony, since that concept is mostly used internally (users shouldn't be dealing directly with seeds).

the `z_exportwallet` RPC call. This confirmation can be performed manually using the

> `zcashd-wallet-tool` utility that is supplied with this release (built or installed in the

Reviewer: Is there any separate doc for the wallet tool or are we relying on the help text?

Author: Just the help text at present.

Reviewer 2: Also note that `zcashd-wallet-tool` doesn't require arguments if you're using the default settings for zcashd's datadir path, port, and RPC authentication. If you've changed those and run it with no arguments, it will tell you what is wrong.

same directory as `zcashd`).  The wallet will not allow the generation of new addresses
until this confirmation has been performed. It is recommended that after this upgrade,

> that funds tied to preexisting addresses be migrated to newly generated addresses so that

Reviewer 2: ```suggestion
funds tied to preexisting addresses be migrated to newly generated addresses so that
```

all wallet funds are recoverable using the emergency recovery phrase going forward. If you
choose not to migrate funds in this fashion, you will continue to need to securely back up
the entire `wallet.dat` file to ensure that you do not lose access to existing funds;
EXISTING FUNDS WILL NOT BE RECOVERABLE USING THE EMERGENCY RECOVERY PHRASE UNLESS THEY
HAVE BEEN MOVED TO A NEWLY GENERATED ADDRESS FOLLOWING THE 4.7.0 UPGRADE.

In the case that your wallet previously contained a Sapling HD seed, the mnemonic phrase
is constructed using the bytes of that seed, such that it is possible to reconstruct 
keys generated using that legacy seed if you know the mnemonic phrase. HOWEVER, THIS 
RECONSTRUCTION DOES NOT FOLLOW THE NORMAL PROCESS OF DERIVATION FROM THE MNEMONIC PHRASE.
Instead, to recover a legacy Sapling key from the mnemonic seed, it is necessary to 
reconstruct the bytes of the legacy seed by conversion of the mnemonic phrase back to

> its source randomness instead of by hashing as is specified in BIP 39. Only keys and

Author: The additional comma here is unnecessary, and would be in conflict with the commas surronding "to recover a legacy Sapling key..." as it would make an additional subordinate clause.

> addresses produced after the upgrade can be obtained by normal ZIP 32 or BIP 39 derivation.

Reviewer 2: ```suggestion
addresses produced after the upgrade can be obtained by normal derivation using BIP 39,
and ZIP 32 (shielded) or BIPs 32 and 44 (transparent).
```

upgrading an existing Zcashd wallet, it is recommended that the wallet be backed up prior
to upgrading to the 4.7.0 Zcashd release.

Following the upgrade to 4.7.0, Zcashd will require that the user confirm that they have
backed up their new emergency recovery phrase, which may be obtained from the output of
the `z_exportwallet` RPC call. This confirmation can be performed manually using the
`zcashd-wallet-tool` utility that is supplied with this release (built or installed in the
same directory as `zcashd`).  The wallet will not allow the generation of new addresses
until this confirmation has been performed. It is recommended that after this upgrade,
funds tied to preexisting addresses be migrated to newly generated addresses so that
all wallet funds are recoverable using the emergency recovery phrase going forward. If you
choose not to migrate funds in this fashion, you will continue to need to securely back up
the entire `wallet.dat` file to ensure that you do not lose access to existing funds;
EXISTING FUNDS WILL NOT BE RECOVERABLE USING THE EMERGENCY RECOVERY PHRASE UNLESS THEY
HAVE BEEN MOVED TO A NEWLY GENERATED ADDRESS FOLLOWING THE 4.7.0 UPGRADE.

> In the case that your wallet previously contained a Sapling HD seed, the mnemonic phrase
> is constructed using the bytes of that seed, such that it is possible to reconstruct 
> keys generated using that legacy seed if you know the mnemonic phrase. HOWEVER, THIS 
> RECONSTRUCTION DOES NOT FOLLOW THE NORMAL PROCESS OF DERIVATION FROM THE MNEMONIC PHRASE.

Reviewer 2: ```suggestion
In the case that your wallet previously contained a Sapling HD seed, the emergency recovery
phrase is constructed using the bytes of that seed, such that it is possible to reconstruct
keys generated using that legacy seed if you know the emergency recovery phrase.
HOWEVER, THIS RECONSTRUCTION DOES NOT FOLLOW THE NORMAL PROCESS OF DERIVATION FROM
THE EMERGENCY RECOVERY PHRASE.
```

> Instead, to recover a legacy Sapling key from the mnemonic seed, it is necessary to 
> reconstruct the bytes of the legacy seed by conversion of the mnemonic phrase back to
> its source randomness instead of by hashing as is specified in BIP 39. Only keys and

Reviewer 2: ```suggestion
Instead, to recover a legacy Sapling key from the emergency recovery phrase, it is necessary
to  reconstruct the bytes of the legacy seed by conversion of the emergency recovery phrase
back to its source randomness instead of by hashing as is specified in BIP 39. Only keys and
```
Unlike the replacements above, the first replacement here is an issue of correctness rather than just tidying up terminology. You *can't* recover the legacy seed from the mnemonic seed, because it's derived using a one-way derivation function. You *can* recover it from the emergency recovery phrase, because (in this case) that's a bijective mapping.

`zcashd-wallet-tool` utility that is supplied with this release (built or installed in the
same directory as `zcashd`).  The wallet will not allow the generation of new addresses
until this confirmation has been performed. It is recommended that after this upgrade,
funds tied to preexisting addresses be migrated to newly generated addresses so that all
wallet funds are recoverable using the emergency recovery phrase going forward. If you
choose not to migrate funds in this fashion, you will continue to need to securely back up
the entire `wallet.dat` file to ensure that you do not lose access to existing funds;
EXISTING FUNDS WILL NOT BE RECOVERABLE USING THE EMERGENCY RECOVERY PHRASE UNLESS THEY
HAVE BEEN MOVED TO A NEWLY GENERATED ADDRESS FOLLOWING THE 4.7.0 UPGRADE.

In the case that your wallet previously contained a Sapling HD seed, the emergency
recovery phrase is constructed using the bytes of that seed, such that it is possible to
reconstruct keys generated using that legacy seed if you know the emergency recovery
phrase. HOWEVER, THIS RECONSTRUCTION DOES NOT FOLLOW THE NORMAL PROCESS OF DERIVATION FROM

> THE MNEMONIC PHRASE.  Instead, to recover a legacy Sapling key from the emergency recovery

Reviewer 2: There's still a use of "mnemonic phrase" here.

New RPC Methods
---------------

> - `walletconfirmbackup` This newly created API checks a provided emergency recovery phrase 

Reviewer: If we don't want people to use this, is it OK to document it here?

Author: It's necessary for full node wallet developers to have this information in the changelog.

  against the wallet's emergency recovery phrase; if the phrases match then it updates the
  wallet state to allow the generation of new addresses.  This backup confirmation

>   workflow can be disabled by starting zcashd with `-requirewalletbackup=false` but this

Reviewer 3: in `rpcwallet.cpp` (in both `master` and `nu5-consensus` https://github.com/zcash/zcash/blob/master/src/wallet/rpcwallet.cpp#L93 / https://github.com/zcash/zcash/blob/nu5-consensus/src/wallet/rpcwallet.cpp#L93) we have `if (GetBoolArg("-walletrequirebackup", params.RequireWalletBackup()) && !pwalletMain->MnemonicVerified())` -- not `requirewalletbackup`

Reviewer 2: ```suggestion
  workflow can be disabled by starting zcashd with `-walletrequirebackup=false` but this
```

>   workflow can be disabled by starting zcashd with `-requirewalletbackup=false` but this

Reviewer 3: ```suggestion
  workflow can be disabled by starting zcashd with `-walletrequirebackup=false` but this
```

Reviewer 2: This should be blocking.

  is not recommended unless you know what you're doing (and have otherwise backed up the
  wallet's recovery phrase anyway).  For security reasons, this RPC method is not intended

>   for use via zcash-cli but is provided to enable `zcashd-wallet-tool` and other

Reviewer 2: ```suggestion
  for use via `zcash-cli` but is provided to enable `zcashd-wallet-tool` and other
```

  third-party wallet interfaces to satisfy the backup confirmation requirement. Use of the

>   `walletconfirmbackup` API via zcash-cli would risk that the recovery phrase being

Reviewer 2: ```suggestion
  `walletconfirmbackup` API via `zcash-cli` would risk that the recovery phrase being
```

  confirmed might be leaked via the user's shell history or the system process table;
  `zcashd-wallet-tool` is specifically provided to avoid this problem.
- `z_getnewaccount` This API allows for creation of new BIP 44 / ZIP 32 accounts using HD
  derivation from the wallet's mnemonic seed. Each account represents a separate spending
  authority and source of funds. A single account may contain funds in the Sapling and
  Orchard shielded pools, as well as funds held in transparent addresses.
- `z_getaddressforaccount` This API allows for creation of diversified unified addresses 
  under a single account. Each call to this API will, by default, create a new diversified

>   unified address containing p2pkh, Sapling, and Orchard receivers.  Additional arguments

Reviewer: Should "p2pkh" be "transparent" for clarity?

Author: ```suggestion
  unified address containing transparent p2pkh, Sapling, and Orchard receivers.  Additional arguments
```

  to this API may be provided to request the address to be created with a user-specified
  set of receiver types and diversifier index.
- `z_getbalanceforaccount` This API makes it possible to obtain balance information on a 
  per-account basis.

> - `z_getbalanceforviewingkey` This newly created API allows a user to obtain balance

Reviewer: Maybe change to "This API..." to be consistent.

Author: ```suggestion
- `z_getbalanceforviewingkey` This API allows a user to obtain balance
```

  information for funds visible to a Sapling or Unified full viewing key; if a Sprout
  viewing key is provided, this method allows retrieval of the balance only in the case
  that the wallet controls the corresponding spending key.  This API has been added to
  supplement (and largely supplant) `z_getbalance`. Querying for balance by a single
  address returns only the amount received by that address, and omits value sent to other
  diversified addresses derived from the same full viewing key; by using
  `z_getbalanceforviewingkey` it is possible to obtain a correct balance that includes all
  amounts controlled by a single spending key, including both those sent to external
  diversified addresses and to wallet-internal change addresses.
- `z_listunifiedreceivers` This API allows the caller to extract the individual component
  receivers from a unified address. This is useful if one needs to provide a bare Sapling

>   or p2pkh address to a service that does not yet support unified addresses.

Reviewer: See earlier comment on p2pkh.

Author: ```suggestion
  or transparent p2pkh address to a service that does not yet support unified addresses.
```

> RPC Changes

Reviewer: I didn't see the update to `z_gettreestate` to include Orchard info. I probably should have updated the release notes with my PR that added this.

Author: Ah, yeah, I suspect we missed that because the change happened on the `master` branch rather than on `nu5-consensus`

New RPC Methods
---------------

- `walletconfirmbackup` This newly created API checks a provided emergency recovery phrase 
  against the wallet's emergency recovery phrase; if the phrases match then it updates the
  wallet state to allow the generation of new addresses.  This backup confirmation
  workflow can be disabled by starting zcashd with `-requirewalletbackup=false` but this
  is not recommended unless you know what you're doing (and have otherwise backed up the
  wallet's recovery phrase anyway).  For security reasons, this RPC method is not intended
  for use via `zcash-cli` but is provided to enable `zcashd-wallet-tool` and other
  third-party wallet interfaces to satisfy the backup confirmation requirement. Use of the

>   `walletconfirmbackup` API via `zcash-cli` would risk that the recovery phrase being

Reviewer 2: For consistency, use "emergency recovery phrase" here also (twice).

  confirmed might be leaked via the user's shell history or the system process table;

> `zcashd-wallet-tool` is provided specifically to avoid this problem.

Reviewer 2: ```suggestion
`zcashd-wallet-tool` is provided specifically to avoid this problem.
```

  transparent addresses when available.
- The result value for `z_importviewingkey` now includes an `address_type` field
  that replaces the now-deprecated `type` key.
- `z_listunspent` has been updated to render unified addresses for Sapling and 
  Orchard outputs when those outputs are controlled by unified spending keys.
  Outputs received by unified internal addresses do not include the `address` field.
- Legacy transparent address generation using `getnewaddress` no longer uses a 
  preallocated keypool, but instead performs HD derivation from the wallet's mnemonic
  seed according to BIP 39 and BIP 44 under account ID `0x7FFFFFFF`.

> ### 'z_sendmany'

Author: Backticks don't behave well in header text.

> - The `z_sendmany` RPC call no longer permits Sprout recipients in the list of recipient

Author: Then you have the same problem with `addresses`. Also, in general we're trying to shift our nomenclature to refer to "receivers" and "recipients" instead of "addresses" which, with UAs, are a less precise concept.

- `z_gettotalbalance` has been deprecated. Use `z_getbalanceforaccount` instead.
- `dumpwallet` has been deprecated. Use `z_exportwallet` instead.

Build System
------------

- Clang has been updated to use LLVM 13.0.1.

> - libcxx has been updated to use LLVM 13.0.1, except on Windows where it uses 13.0.0-3.

Reviewer 2: ```suggestion
- libc++ has been updated to use LLVM 13.0.1, except on Windows where it uses 13.0.0-3.
```
It's occasionally called `libcxx` (and [somewhat inconsistently](https://github.com/zcash/zcash/blob/master/depends/packages/libcxx.mk#L26-L27)) in filenames, [etc.](https://lists.llvm.org/mailman/listinfo/libcxx-commits), but the official name is libc++.

Author: I will make this correction in the release PR.

- The Rust toolchain dependency has been updated to version 1.59.0.

> Platform Support

Reviewer: We're adding Bullseye support I believe so we should include that platform.

----------------

> - Debian 9 has been removed from the list of supported platforms.

Author: ```suggestion
- Debian 9 has been removed from the list of supported platforms.
- Debian 11 (Bullseye) has been added to the list of supported platforms.
```

- A build issue (a missing header file) has been fixed for macOS targets.

> - On Arch Linux only, use Debian's libtinfo5_6.0 to fix a build regression.

Reviewer 2: ```suggestion
- On Arch Linux only, Debian's libtinfo5_6.0 is used to fix a build regression.
```

Mining
------

> - Mining to Orchard recipients is now supported on testnet. 

Reviewer: recipients -> receivers? Or "Mining to a Unified Address that contains an Orchard receiver" possibly.

Author: The reason I used "recipients" here is that, although the only way to construct an Orchard recipient address is a unified address containing an Orchard receiver, that's an implementation detail.
